[BE] 이근표 👥 오늘의 짝꿍은?

rootTiket opened the pull request:
## 테스트 결과
---
<img width="1322" alt="스크린샷 2024-04-11 오후 6 52 22" src="https://github.com/Leets-Official/leets-mate-java-1/assets/92284769/6ab37f3e-e1c8-4e66-903a-43950e973378">

## 실행결과
---
<img width="432" alt="스크린샷 2024-04-11 오후 6 54 36" src="https://github.com/Leets-Official/leets-mate-java-1/assets/92284769/922e7611-9b1f-46e5-9126-f8308ea37075">

## 기능 명세서
---
✅ 입력받은 이름을 , 기준으로 이름 리스트 생성
✅ 입력된 이름 수를 저장하는 메서드 구현
✅ 짝수를 입력 받고 리스트의 크기 보다 크면 예외를 처리
✅ 다시 뽑기 실행시 y,n 을 제외한 문자가 입력 될 경우 예외 처리
✅ 짝을 뽑고 주어진 형식에 맞게 출력하는 메서드 구현

rootTiket committed:
[feat]: 짝 뽑기 기능 구현

taeseokyang commented on src/main/java/leets/leets_mate/LeetsMateApplication.java:
```diff
@@ -6,33 +6,107 @@ public class LeetsMateApplication {
     public void run() {
+        System.out.println("멤버의 이름을 입력해 주세요. (, 로 구분)");
+        Scanner sc = new Scanner(System.in);
+        String input = sc.nextLine();
```
무슨 입력인지 나타내는 변수명이면 좋을 것 같습니다!

taeseokyang commented on src/main/java/leets/leets_mate/LeetsMateApplication.java:
```diff
@@ -6,33 +6,107 @@ public class LeetsMateApplication {
+        System.out.println("최대 짝 수를 입력해 주세요.");
+        int maxNum = Integer.parseInt(sc.nextLine());
+        checkDataValidity(memberList.size(), maxNum);
+        String flag ="true";
```
무슨 플래그 인지 알려주면 좋을거 같습니다!
변수명만 보고 바로 이해 하면 코드를 읽기 더욱 편할 거 같습니다!

boolean 형 변경하는 것도 좋다 생각합니다! 만약 그렇다면 isDone은 어떨까요?
현재 String 형으로 true, exception을 반환 받고 있는데, isContinue() 함수에서는 계속 할지 안할지 true, false 둘중 하나만 반환하면 된다 생각합니다!

taeseokyang commented on src/main/java/leets/leets_mate/LeetsMateApplication.java:
```diff
@@ -6,33 +6,107 @@ public class LeetsMateApplication {
     // 문자열로된 멤버들을 리스트로 분리하는 함수입니다.
     public List<String> parseMembers(String members) {
-        return new ArrayList<>();
+        List<String> memberList = new ArrayList<>();
```
`return Arrays.asList(members.split(","));`
위와 같이 작성 할 수 있습니다!

taeseokyang reviewed (commented):
수고하셨습니다!!!!

ay-eonii commented on src/main/java/leets/leets_mate/LeetsMateApplication.java:
```diff
@@ -6,33 +6,107 @@ public class LeetsMateApplication {
+        int maxNum = Integer.parseInt(sc.nextLine());
+        checkDataValidity(memberList.size(), maxNum);
+        String flag ="true";
+        while (flag.equals("true") || flag.equals("exception")) {
```
만약 flag 가 null 이라면 어떤 문제가 있을까요?
그리고 이를 방지하기 위해 어떻게 해야할까요?

ay-eonii commented on src/main/java/leets/leets_mate/LeetsMateApplication.java:
```diff
@@ -6,33 +6,107 @@ public class LeetsMateApplication {
     public void checkHasNoEnglish(String members) {
+        try {
+            for (char c : members.toCharArray()) {
+                if (!(c >= '가' && c <= '힣') && !(Character.toString(c).equals(","))) {
```
복잡한 조건문은 메서드로 따로 빼면 가독성이 죠아져요 👍

ay-eonii commented on src/main/java/leets/leets_mate/LeetsMateApplication.java:
```diff
@@ -6,33 +6,107 @@ public class LeetsMateApplication {
+            }
+        } catch (Exception e) {
+            System.out.println(e.getMessage());
+            run();
```
숫자 실수하면 이름까지 다시 입력받는 하드코어 프로그램이 되었네요! 👍

ay-eonii commented on src/main/java/leets/leets_mate/LeetsMateApplication.java:
```diff
@@ -6,33 +6,107 @@ public class LeetsMateApplication {
     // 결과를 프린트 하는 함수입니다.
     public void printResult(List<List<String>> result) {
+        System.out.println("오늘의 짝 추천 결과입니다.");
+        for (List<String> line : result) {
+            StringBuilder sb = new StringBuilder();
+            sb.append("[ ");
+            for (String s : line) {
+                sb.append(s).append(" | ");
+            }
+            sb.replace(sb.length()-2,sb.length(),"]");
+            System.out.println(sb);
+        }
```
StringJoiner에 대해 알아보면 편해질거에요!

ay-eonii commented on src/main/java/leets/leets_mate/LeetsMateApplication.java:
```diff
@@ -6,33 +6,107 @@ public class LeetsMateApplication {
+        System.out.print("다시 구성하시겠습니까? (y or n): ");
+        Scanner sc = new Scanner(System.in);
+        String q = sc.nextLine();
+        if (q.equals("n")) {
+            System.out.println("자리를 이동해 서로에게 인사해주세요.");
+            return "false";
+        } else if (q.equals("y")) {
+            System.out.println("--------------------------------");
+            return "true";
+        } else {
+            System.out.println("[ERROR] y 혹은 n만 입력해주세요.");
+            return "exception";
+        }
```
else 는 없어도 되겠네요!

ay-eonii commented on src/main/java/leets/leets_mate/LeetsMateApplication.java:
```diff
@@ -6,33 +6,107 @@ public class LeetsMateApplication {
+    public static String isContinue() {
+        System.out.print("다시 구성하시겠습니까? (y or n): ");
+        Scanner sc = new Scanner(System.in);
+        String q = sc.nextLine();
```
`q` 는 무엇을 의미하나요?

ay-eonii reviewed (commented):
1주차 미션 고생하셨습니다
char로 한글을 판단한게 인상깊네요! 👍

여유가 있다면 클래스가 적절한 책임을 가지도록 분리해봐도 좋겠어요!

rootTiket committed:
[refactor]: 멤버 입력 변수 명 수정

rootTiket committed:
[refactor]: 다시 뽑기 기능 수정

-다시 뽑기를 위한 while 문의 flag 변수 타입을 Boolean으로 변경
-flag 변수의 이름을 isDone으로 수정
-isContinue의 반환값을 Boolean으로 수정

rootTiket committed:
[refactor]: 멤버 리스트 분리 하는 함수 수정
-for문 대신 asList를 이용하여 멤버를 분리하도록 변경

jiixon commented on src/main/java/leets/leets_mate/LeetsMateApplication.java:
```diff
@@ -6,33 +6,105 @@ public class LeetsMateApplication {
+        System.out.println("최대 짝 수를 입력해 주세요.");
+        int maxNum = Integer.parseInt(sc.nextLine());
+        checkDataValidity(memberList.size(), maxNum);
+        Boolean isDone = true;
+        while (isDone .equals(true)) {
+            printResult(generateRandomGroups(memberList,maxNum));
+            isDone = isContinue();
```
`while (isDone) {` 이렇게 구현해도 되지않을까요?

jiixon reviewed (commented):
고생하셨습니다~
